torchci/lib/metricUtils.ts

```diff
@@ -0,0 +1,193 @@
+import { JobsPerCommitData, JobAnnotation } from "lib/types";
+
+// When N consecutive failures of the same type happen, the failures are counted as
+// broken trunk failures (approximately)
+export const BROKEN_TRUNK_THRESHOLD = 3;
+
+// When more than N failures happening in the same commit, the failures are counted
+// as part of an outage or broken infra (approximately)
+export const OUTAGE_THRESHOLD = 10;
+
+function getFailureByJobName(
+  jobName: string,
+  failures: { [jobName: string]: { [t: string]: number } }
+) {
+  if (!(jobName in failures)) {
+    failures[jobName] = {
+      [JobAnnotation.BROKEN_TRUNK]: 0,
+      [JobAnnotation.INFRA_BROKEN]: 0,
+      [JobAnnotation.TEST_FLAKE]: 0,
+    };
+  }
+
+  return failures[jobName];
+}
+
+function updateBrokenInfraCount(
+  jobName: string,
+  count: number,
+  failures: { [jobName: string]: { [t: string]: number } }
+) {
+  if (count === 0) {
+    return;
+  }
+
+  const failure = getFailureByJobName(jobName, failures);
+  failure[JobAnnotation.INFRA_BROKEN] += count;
+}
+
+function updateFailureCount(
+  jobName: string,
+  count: number,
+  failures: { [jobName: string]: { [t: string]: number } },
+  is_broken_trunk: boolean
+) {
+  if (count === 0) {
+    return;
+  }
+
+  const failure = getFailureByJobName(jobName, failures);
+  failure[
+    is_broken_trunk ? JobAnnotation.BROKEN_TRUNK : JobAnnotation.TEST_FLAKE
+  ] += count;
+}
+
+export function approximateSuccessByJobName(
+  // The data from Rockset is sorted by time DESC, so newer commits come first
+  data?: JobsPerCommitData[]
+) {
+  const successesByJobName: { [success: string]: number } = {};
+
+  if (data === undefined || data === null) {
+    return successesByJobName;
+  }
+
+  data.forEach((commit: JobsPerCommitData) => {
+    const successes = new Set(
+      commit.successes.filter(
+        (n) => n !== null && n !== undefined && n.length > 0
+      )
+    );
+
+    // Iterate though all the successes in the commit and aggregate them by name
+    successes.forEach((success: string) => {
+      if (!(success in successesByJobName)) {
+        // Start the count
+        successesByJobName[success] = 0;
```

Comment:
Ah, the comment threw me off :P 

```suggestion
        // Make sure the dict is initialized
        successesByJobName[success] = 0;
```